packages/site/cms-site/src/sitePreview/SitePreviewApiHelper.ts

```diff
@@ -0,0 +1,71 @@
+import { NextApiRequest, NextApiResponse } from "next";
+
+type Scope = Record<string, unknown>;
+type GraphQLClient = {
+    setHeader(key: string, value: string): unknown;
+    request(document: string, variables?: unknown): Promise<{ isAllowedSitePreview: boolean }>;
+};
+export type SitePreviewParams = {
+    scope: Scope;
+    path: string;
+    settings: {
+        includeInvisible: boolean;
+    };
+};
+export type BlockPreviewParams = {
+    scope: Scope;
+    path: string;
+};
+
+export async function getValidatedSitePreviewParams(
+    req: NextApiRequest,
+    res: NextApiResponse,
+    graphQLClient: GraphQLClient,
+): Promise<SitePreviewParams> {
+    return {
+        scope: await getValidatedScope(req, res, graphQLClient),
+        path: req.query.path as string,
+        settings: JSON.parse(req.query.settings as string),
+    };
+}
+
+export async function getValidatedBlockPreviewParams(
+    req: NextApiRequest,
+    res: NextApiResponse,
+    graphQLClient: GraphQLClient,
+): Promise<BlockPreviewParams> {
+    return {
+        scope: await getValidatedScope(req, res, graphQLClient),
+        path: req.query.path as string,
+    };
+}
+
+async function getValidatedScope(req: NextApiRequest, res: NextApiResponse, graphQLClient: GraphQLClient): Promise<Scope> {
+    const scope = JSON.parse(req.query.scope as string);
+    if (!scope) throw new Error("Scope is missing");
+
+    let isAllowed = false;
+    graphQLClient.setHeader("authorization", req.headers["authorization"] || "");
+    try {
+        const { isAllowedSitePreview } = await graphQLClient.request(
+            "query isAllowedSitePreview($scope: JSONObject!) { isAllowedSitePreview(scope: $scope) }",
+            { scope },
+        );
+        if (isAllowedSitePreview) isAllowed = true;
+        // eslint-disable-next-line @typescript-eslint/no-explicit-any
+    } catch (error: any) {
+        const hasUnauthenticatedRespone =
+            Array.isArray(error?.response?.errors) &&
+            error.response.errors.some(
+                // eslint-disable-next-line @typescript-eslint/no-explicit-any
+                (e: any) => e.extensions?.code === "FORBIDDEN",
+            );
+        if (!hasUnauthenticatedRespone) throw error;
```

Comment:
I'm not sure about using the exception to check access here. Wouldn't it be cleaner to return false if a user hasn't access? You could disable the scope check for the resolver and perform the check using the access control service.